File: 0513-find-bottom-left-tree-value/0513-find-bottom-left-tree-value.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def heightOfTree(self, root: Optional[TreeNode]) -> int:
        return max(self.heightOfTree(root.left), self.heightOfTree(root.right))+1 if root else 0
    
    def findBottomLeftValue(self, root: Optional[TreeNode]) -> int:
        
        height = self.heightOfTree(root)
        
        treeque = deque()
        
        treeque.append(root) 
        
        while height>1:
            
            baseLength = len(treeque)
            
            while baseLength:
                currNode = treeque.popleft()
                treeque.append(currNode.left) if currNode.left else None
                treeque.append(currNode.right) if currNode.right else None
                baseLength -= 1
            
            height -=1 
        
        firstNode = treeque.popleft()
        return firstNode.val
                
                
        
        
        
        

File: 0513-find-bottom-left-tree-value/0513-find-bottom-left-tree-value.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    int heightOfTree(TreeNode* root) {
        if(!root)
            return 0;
        return max(heightOfTree(root->left), heightOfTree(root->right))+1;
    }
    
    int findBottomLeftValue(TreeNode* root) {
        int height = heightOfTree(root);
        queue<TreeNode*> qu;
        qu.push(root);
        while(height>1) {
            int n = qu.size();
            while(n--) {
                TreeNode* temp = qu.front();
                qu.pop();
                if(temp->left)
                    qu.push(temp->left);
                if(temp->right)
                    qu.push(temp->right);
            }
            height--;
        }
        return qu.front()->val;
    }
};